libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c

```diff
@@ -355,30 +381,173 @@ AwsIotProvisioningError_t _AwsIotProvisioning_SerializeCreateKeysAndCertificateR
+            status = AWS_IOT_PROVISIONING_INTERNAL_FAILURE;
+        }
+
+        /* Close the map.
+         * Note: Always close the container, even if appendVKeyValue() fails,
+         * to free the memory of the map encoder object.
+         */
+        if( isSuccessStatus( _pAwsIotProvisioningEncoder->closeContainer( pEncoder,
+                                                                          &mapEncoder ) ) == false )
+        {
+            IotLogError( "serializer: Unable to serialize request payload: "
+                         "Failed to close map container: Operation={%s}",
+                         CREATE_CERT_FROM_CSR_OPERATION_LOG );
+            status = AWS_IOT_PROVISIONING_INTERNAL_FAILURE;
+        }
+    }
+
+    return status;
+}
+
+/*------------------------------------------------------------------*/
+
 AwsIotProvisioningError_t _AwsIotProvisioning_CalculateCertFromCsrPayloadSize( const char * pCertificateSigningRequest,
                                                                                size_t csrLength,
                                                                                size_t * pPayloadSize )
 {
-    ( void ) pCertificateSigningRequest;
-    ( void ) csrLength;
-    ( void ) pPayloadSize;
+    AwsIotProvisioning_Assert( pCertificateSigningRequest != NULL );
+    AwsIotProvisioning_Assert( csrLength != 0 );
+    AwsIotProvisioning_Assert( pPayloadSize != NULL );
+
+    bool status = true;
+    IotSerializerEncoderObject_t outerEncoder = IOT_SERIALIZER_ENCODER_CONTAINER_INITIALIZER_STREAM;
+
+    if( _pAwsIotProvisioningEncoder->init( &outerEncoder,
+                                           NULL,
+                                           0 ) != IOT_SERIALIZER_SUCCESS )
+    {
+        IotLogError( "serializer: Unable to serialize request payload: "
+                     "Failed to initialize encoder: Operation={%s}",
+                     CREATE_CERT_FROM_CSR_OPERATION_LOG );
+        status = AWS_IOT_PROVISIONING_INTERNAL_FAILURE;
+    }
+
+    if( status == true )
+    {
+        /* Perform a dry-run serialization of the Certificate-Signing Request */
+        /* data to calculate the size of the payload. */
+        status = _serializeCertFromCsrPayload( pCertificateSigningRequest,
+                                               csrLength,
+                                               &outerEncoder,
+                                               true );
+    }
+
+    if( status == true )
+    {
+        *pPayloadSize = _pAwsIotProvisioningEncoder->getExtraBufferSizeNeeded( &outerEncoder );
+        AwsIotProvisioning_Assert( *pPayloadSize != 0 );
+        IotLogDebug( "serializer: Calculated serialization size and populated in output parameter: "
+                     "Operation={%s}",
+                     CREATE_CERT_FROM_CSR_OPERATION_LOG );
+    }
+
+    _pAwsIotProvisioningEncoder->destroy( &outerEncoder );
 
     return true;
```

Comment:
Why does this function always return `true`? That's not even its return type.

Reply:
Fixed.